### ISSUE TITLE
Pong's CancelledError races with ConnectionClosed error

### ISSUE THREAD

**Issue by lgrahl** [aaugustin/websockets#464]
I upgraded to the current version of this library and I find it difficult to handle `asyncio.CancelledError` in case the pong is being cancelled due to a disconnect.

What happens in my server is that I'm waiting for pongs in one task while sending and receiving in another task. Both tasks are being waited for. The result is that in some occasions, `asyncio.CancelledError` is raised before `websockets.ConnectionClosed` and vice versa. When the `asyncio.CancelledError` is raised first, there is no context on why it has been raised.

Another problem with `asyncio.CancelledError` is that there's no way (IIRC) of finding out whether the current task has been cancelled or the task I'm waiting for has been cancelled. But this mainly is an awkward concept of asyncio and not this libraries fault. It seems one has to be extremely careful on when to cancel a task and when not to cancel a task and instead use other ways to signal cancellation (which highly depends on the task). This may clash with the design choice of this library.

TL;DR: I would prefer it if this library would raise `websockets.ConnectionClosed` in the pong's future instead of cancelling the pong.

**Comment by aaugustin** [comment]
If I understand correctly, the question is:

```python
# may raise websockets.ConnectionClosed
pong_waiter = await ws.ping()

# may raise asyncio.CancelledError; should it raise websockets.ConnectionClosed instead?
await pong_waiter
```

I've been researching when `Future.cancel()` is more or less appropriate than `Future.set_exception()` and not finding much.

There's a minor difficulty for implemnting this change: at the point where pending "ping waiters" are cancelled, the close code and reason may not be known yet, which prevents instantiating the proper `websockets.ConnectionClosed` exception.

Another drawback is that, if there are pending pings when the connection is closed and nothing awaits them, an exception will be logged. Per PEP3156:

> When a Future is garbage-collected, if it has an associated exception but neither result() nor exception() has ever been called, the exception is logged.

**Comment by lgrahl** [comment]
> I've been researching when `Future.cancel()` is more or less appropriate than `Future.set_exception()` and not finding much.

Same problem here. It seems there is no consensus on what's best practice when it comes to cancellation of futures. No one did respond in the Python IRC channel either.

I've tinkered around all day, trying to grasp the concept of `asyncio.CancelledError` a little bit further. My current impression is that cancelling futures is okay if you really don't care about it any longer and it can be discarded (e.g. when cleaning up a connection). Which is why I think that futures, which have been created via a public API call, should not be cancelled by anyone else but the caller. I base this on the fact that only the caller (and thus the creator of the future) knows when the future will be discarded and never be waited for again.

I don't know if there should ever be a use case for catching a `asyncio.CancelledError` other than for cleanup purposes since one is absolutely clueless on where the cancellation happened (deep inside the call stack or from the outside).
In the case of `pong`: If one catches an `asyncio.CancelledError` inside of a task, it is unknown whether the task who is awaiting the pong has been cancelled or the pong itself has been cancelled. One may want to apply special handling for these two cases but differentiating between them is near impossible.

My take on the cancellation concept may be completely wrong though... perhaps we should ping one of the people who designed asyncio (and then look at how it could be resolved).

**Comment by aaugustin** [comment]
Cancellation is interesting for a Task: it's the only way you can "inject an exception" and terminate the Task from the outside.

I find it less interesting for a Future: you can always inject an exception with `Future.set_exception()`.

----

Eventually I found this post by GvR that confirms your intuition: https://groups.google.com/d/msg/python-tulip/LZQe38CR3bg/7qZ1p_q5yycJ

> A cancellation, however, is different. It is initiated by the *consumer* to indicate that it is no longer interested in the result.

This sentence means the same thing than when you said:

> futures, which have been created via a public API call, should not be cancelled by anyone else but the caller.

So we should make the change you're suggesting. (And also review other cancellations, in case the same problem exists somewhere else in the library.)

Also note that `ping` uses `asyncio.shield` in order to keep the "internal" future alive even if something cancelled the future it sent into the outside world. This is necessary for websockets to do proper bookkeeping of pings and pongs.

Finally I believe that no exceptions should be logged as a result of this change if `WebSocketCommonProtocol.keepalive_ping_task` is cancelled before injecting the exceptions.

**Comment by aaugustin** [comment]
Since the current behavior was publicly documented in the changelog (see #379), this will be a backwards incompatible change.

**Comment by lgrahl** [comment]
Thanks for researching. :+1: Guido put it into better words. And I fully concur with your plan on how to resolve it.

**Comment by aaugustin** [comment]
@lgrahl Would you mind reviewing #467?

Is it possible to deploy the change in an environment where you're seeing the issue and confirm that it's resolved?

**Comment by lgrahl** [comment]
Yep, I'll try that in the upcoming week.
